crates/autopilot/src/limit_orders/quoter.rs

```diff
@@ -156,16 +169,146 @@ impl LimitOrderQuoter {
     }
 }
 
+fn query_from(data: &OrderQuotingData) -> Query {
+    Query {
+        owner: H160(data.owner.0),
+        token: H160(data.sell_token.0),
+        source: sell_token_source_from(data.sell_token_balance),
+    }
+}
+
+fn order_spec_from(data: OrderQuotingData) -> OrderFeeSpecifier {
+    OrderFeeSpecifier {
+        sell_token: data.sell_token,
+        buy_token: data.buy_token,
+        sell_amount: data.sell_amount,
+    }
+}
+
+async fn orders_with_sufficient_balance(
+    balance_fetcher: &dyn BalanceFetching,
+    mut orders: Vec<OrderQuotingData>,
+) -> Vec<OrderQuotingData> {
+    let queries = orders.iter().map(query_from).unique().collect_vec();
+    let balances = balance_fetcher.get_balances(&queries).await;
+    let balances: HashMap<_, _> = queries
+        .iter()
+        .zip(balances.into_iter())
+        .filter_map(|(query, result)| Some((query, result.ok()?)))
+        .collect();
+
+    let order_count_before = orders.len();
+
+    orders.retain(|order| {
+        if let Some(balance) = balances.get(&query_from(order)) {
+            let has_sufficient_balance = match order.partially_fillable {
+                true => balance >= &U256::one(), // any amount would be enough
+                false => balance >= &big_decimal_to_u256(&order.sell_amount).unwrap(),
+            };
+            return has_sufficient_balance;
+        }
+
+        // In case the balance couldn't be fetched err on the safe side and assume
+        // the order can be filled to not discard limit orders unjustly.
+        true
+    });
+
+    Metrics::get()
+        .orders_skipped_for_missing_balance
+        .set((order_count_before - orders.len()) as i64);
+    orders
+}
+
+#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq, clap::ValueEnum)]
+#[clap(rename_all = "verbatim")]
+pub enum QuotingStrategy {
+    SkipUnfunded,
+    // TODO add `PrioritizeByPrice`
+}
+
 #[derive(prometheus_metric_storage::MetricStorage, Clone, Debug)]
 #[metric(subsystem = "limit_order_quoter")]
 struct Metrics {
     /// Categorizes order quote update results.
     #[metric(labels("type"))]
     update_result: prometheus::IntCounterVec,
+
+    /// Tracks how many orders don't get quoted because their
+    /// owners don't have sufficient balance.
+    orders_skipped_for_missing_balance: prometheus::IntGauge,
```

Comment:
Oh wow! the `#[metrics()]` really isn't needed then 🤔.

Reply:
It's only needed if you want to add some configuration to that metric. For example bucket sizes or labels. Since the defaults are fine we don't need to use the macro here.